iac-extensions/arm/src/test/java/org/sonar/iac/arm/visitors/ArmSymbolVisitorTest.java

```diff
@@ -0,0 +1,459 @@
+
+    Symbol symbol = symbolTable.getSymbol("foo");
+    assertThat(symbol).isNotNull();
+
+    assertThat(symbol.usages()).hasSize(2);
+    assertThat(symbol.usages())
+      .filteredOn(usage -> usage.kind() == Usage.Kind.ACCESS)
+      .hasSize(1)
+      .allSatisfy(usage -> assertThat(usage.tree().getKind()).isEqualTo(ArmTree.Kind.VARIABLE));
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+    // no access usage should be registered when same name as variable in outputDeclaration
+    "bicep," + "output foo string =  deployment().name",
+    "bicep," + "output foo string =  baba['foo']",
+
+    // JSON
+    "json," + "[deployment().name]",
+    "json," + "[baba['foo']]"
+  })
+  void shouldRegisterNoUsageAccess(String language, String codeStatement) {
+    String code = ArmSourceCodeBuilder.create(language)
+      .addVariableDeclaration(VARIABLE_DECLARATION.get(language))
+      .addCodeStatement(OUT, codeStatement)
+      .build();
+
+    File file = scanFile(code);
+
+    Statement variableDeclaration = file.statements().stream()
+      .filter(s -> s instanceof VariableDeclaration)
+      .findFirst()
+      .orElseThrow();
+    SymbolTable symbolTable = file.symbolTable();
+
+    assertThat(symbolTable).isNotNull();
+    assertThat(symbolTable.hasFoundUnresolvableVariableAccess()).isFalse();
+    assertThat(symbolTable.getSymbols()).hasSize(1);
+
+    Symbol symbol = symbolTable.getSymbol("foo");
+    assertThat(symbol).isNotNull();
+
+    assertThat(symbol.usages()).hasSize(1).allSatisfy(usage -> {
+      assertThat(usage.kind()).isEqualTo(Usage.Kind.ASSIGNMENT);
+      assertThat(usage.tree()).isEqualTo(variableDeclaration);
+    });
+  }
+
+  @ParameterizedTest
+  @MethodSource("languagesToTest")
+  void shouldRegisterUsageWhenDeclarationAfterAccess(String language) {
+    String code = ArmSourceCodeBuilder.create(language)
+      .addVariableDeclaration(VARIABLE_DECLARATION_WITH_USAGE.get(language))
+      .addVariableDeclaration(VARIABLE_DECLARATION.get(language))
+      .build();
+
+    File file = scanFile(code);
+
+    SymbolTable symbolTable = file.symbolTable();
+
+    assertThat(symbolTable).isNotNull();
+    assertThat(symbolTable.hasFoundUnresolvableVariableAccess()).isFalse();
+    assertThat(symbolTable.getSymbols()).hasSize(2);
+
+    Symbol symbol = symbolTable.getSymbol("foo");
+    assertThat(symbol).isNotNull();
+
+    assertThat(symbol.usages()).hasSize(2);
+    assertThat(symbol.usages())
+      .filteredOn(usage -> usage.kind() == Usage.Kind.ACCESS)
+      .hasSize(1);
+  }
+
+  @ParameterizedTest
+  @MethodSource("languagesToTest")
+  void shouldOnlyCreateOneAccessUsageWhenRegisteringIdentifierMultipleTimes(String language) {
+    String code = ArmSourceCodeBuilder.create(language)
+      .addVariableDeclaration(VARIABLE_DECLARATION.get(language))
+      .addVariableDeclaration(VARIABLE_DECLARATION_WITH_USAGE.get(language))
+      .build();
```

Comment:
I have mixed feelings about the `ArmSourceCodeBuilder`. From one side it is a nice generic API, from the other side it is a test code and it is harder to imagine what will be the content of `code` after executing these few lines. 
It's ok to leave it as it is but maybe next time it is better to choose simpler approach.